### PR TITLE
Adds audio player in audio widget

### LINKED ISSUE #2028: Improve audio widget
#### Problem description
Clicking on play button in `Audio Widget` opens a new app to play the audio file. Instead, an audio player should be embedded within the widget to play and control the audio.

#### Steps to reproduce the problem
1. Open any form containing audio widget.
2. Choose any audio file 
3. Click on play button 

#### Expected behavior
If the audio file is available then a player should appear within the audio widget

#### Proposed solution
image

@opendatakit-bot claim

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Closes #2028 

Related to #1120 (previous issue),  #1127 #1145 

#### What has been done to verify that this works as intended?
Tried running on Redmi Note 4 and Moto G5 Plus (Android 7.0) and found the behavior correct

#### Why is this the best possible solution? Were any other approaches considered?
This was the only approach considered

#### Are there any risks to merging this code? If so, what are they?
Yes, the audio widget's functionality might get affected so it needs testing. 

#### Do we need any specific form for testing your changes? If so, please attach one.
`All Widget Form` or any other form containing `Audio Widget`

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+     * Seeks media to the new position and updates timer
+     */
+    private void seekTo(int newPosition) {
+        if (newPosition < 0) {
+            newPosition = 0;
+        } else if (newPosition > mediaPlayer.getDuration()) {
+            newPosition = mediaPlayer.getDuration();
+        }
+
+        mediaPlayer.seekTo(newPosition);
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: Seven lines can be one, with our friends min and max.

screen shot 2018-03-18 at 8 39 42 pm

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    };
+
+    private State state;
+    private FormEntryPrompt formEntryPrompt;
+    private View view;
+
+    AudioController() {
+    }
+
+    void init(Context context, MediaPlayer mediaPlayer, FormEntryPrompt formEntryPrompt) {
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: I wish the fields could be final and set in the constructor. Do you need this `init` for testing?

File: collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java [hunk 3]
```diff
@@ -125,4 +125,29 @@ public static DatePickerDetails getDatePickerDetails(String appearance) {
 
         return new DatePickerDetails(datePickerType, datePickerMode);
     }
+
+    /**
+     * Function to convert milliseconds time to Hours:Minutes:Seconds Format
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: May I suggest,
Converts milliseconds time to Hours:Minutes:Seconds format

File: collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java [hunk 4]
```diff
@@ -125,4 +125,29 @@ public static DatePickerDetails getDatePickerDetails(String appearance) {
...
+     */
+    public static String millisToTimer(long milliseconds) {
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: Did you consider Joda-Time, or whatever the modern equivalent might be?

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    SeekBar seekBar;
+
+    private Context context;
+    private MediaPlayer mediaPlayer;
+    private Handler seekHandler = new Handler();
+
+    /**
+     * Background Runnable thread
+     */
+    private Runnable updateTimeTask = new Runnable() {
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: Making this `final` may ease worries that it changes later.

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java (hunk 2 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: Only context can be set in the constructor as `mediaPlayer` and `formEntryPrompt` can only be set once the super constructor has been called

Comment (dcbriccetti) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: OK. Thanks.

File: collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java (hunk 4 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: No, I don't know about that. Can you please guide me how to do that?

Comment (dcbriccetti) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: Something like—or perhaps exactly—this:
```
public static String millisToTimer(long milliseconds) {
    DateTimeFormatter fmt = DateTimeFormat.shortTime();
    return fmt.print(new DateTime(milliseconds));
}
```

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+            newPosition = mediaPlayer.getDuration();
+        }
+
+        mediaPlayer.seekTo(newPosition);
+        updateTimer();
+    }
+
+    @OnClick(R.id.playBtn)
+    void playClicked() {
+        state = mediaPlayer.isPlaying() ? pause() : play();
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: I don’t like that pause and play unconditionally return the PAUSED and PLAYING states, respectively, but it does make this line of code nice and clean.

**Review by dcbriccetti — APPROVED** [review]

File: collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java (hunk 6 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: Does this look good?
```
@OnClick(R.id.playBtn)
    void playClicked() {
        if (mediaPlayer.isPlaying()) {
            pause();
            state = State.PAUSED;
        } else {
            play();
            state = State.PLAYING;
        }

    }
```

Comment (dcbriccetti) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: Yes, and if you go that way, you can make pause and play return void.

Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java]: :+1:

**Comment by mmarciniak90** [comment]
**Tested with success!**

Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0

Verified cases:
* recorded audio
* added existed file
* removed file
* new and saved form

screenshot_20180323-134536

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by grzesiek2010** [comment]
@shobhitagarwal1612 just a small thing: wouldn't it be better to use the same margin size like for buttons?

**Comment by akshay-ap** [comment]
I think the buttons to controls the audio player are placed too close. Users with small screen size and relatively bigger finger might face difficulty. It would be better to separate the buttons equally and use complete horizontal screen space.

**Comment by shobhitagarwal1612** [comment]
> just a small thing: wouldn't it be better to use the same margin size like for buttons

Sure

**Comment by shobhitagarwal1612** [comment]
image

@akshay-ap @grzesiek2010  
Fixed

File: collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java [hunk 7]
```diff
@@ -125,4 +125,29 @@ public static DatePickerDetails getDatePickerDetails(String appearance) {
 
         return new DatePickerDetails(datePickerType, datePickerMode);
     }
+
+    /**
+     * Converts milliseconds time to Hours:Minutes:Seconds format
+     */
+    public static String millisToTimer(long milliseconds) {
+        StringBuilder sb = new StringBuilder();
+
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: I suggest you remove this method and have the caller use DateTimeFormatter instead.

**Review by dcbriccetti — CHANGES_REQUESTED** [review]
Sorry for repeating myself. I want to make sure Github knows I have requested this.

File: collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java (hunk 7 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: @dcbriccetti  Done. 
I used `new DateTime(millis, DateTimeZone.UTC).toString("mm:ss");` to convert `millis` to `mm:ss`

Comment (dcbriccetti) [reply collect_app/src/main/java/org/odk/collect/android/utilities/DateTimeUtils.java]: Oh, even better than what I was thinking!

**Comment by lognaturel** [comment]
> Users with small screen size and relatively bigger finger might face difficulty.

Really important consideration, thanks for bringing it up, @akshay-ap! With the latest change it feels easy to hit and about the same touch target size as my built-in player. Let's specifically ask for feedback on this in the beta.

@dcbriccetti will let you do the final approval and merge!

**Review by lognaturel — APPROVED** [review]

**Review by dcbriccetti — APPROVED** [review]

**Pull request merged**